src/v/kafka/client/direct_consumer/fetcher.cc

```diff
@@ -10,36 +10,133 @@
+  model::node_id id, prefix_logger& logger)
+  : _id(id)
+  , _logger(&logger)
+  , _fetch_sessions_enabled(enable_fetch_sessions.bind()) {
+    reset();
+    _fetch_sessions_enabled.watch(
+      [this]() { toggle(_fetch_sessions_enabled()); });
+}
+
+void fetch_session_state::update_fetch_session(kafka::fetch_session_id id) {
+    auto deferred_log = ss::defer(
+      [this, prev_ss = session_state, prev_id = session_id] {
+          if (prev_id == session_id && prev_ss == session_state) {
+              return;
+          }
+          vlog(
+            _logger->trace,
+            "[broker: {}] {{session {}: {}}} -> {{session {}: {}}}",
+            _id,
+            prev_id,
+            prev_ss,
+            session_id,
+            session_state);
+      });
+    switch (session_state) {
+    case state::none:
+        reset();
+        return;
+    case state::need_full_fetch:
+        reset();
+        if (id != kafka::invalid_fetch_session_id) {
+            session_id = id;
+            session_state = state::incremental_fetch;
+            advance_epoch();
+        }
+        return;
+    case state::incremental_fetch:
+        if (id == kafka::invalid_fetch_session_id) {
+            reset();
+        } else if (id == session_id) {
+            session_state = state::incremental_fetch;
+            advance_epoch();
+        } else {
+            close();
+        }
+        return;
+    case state::needs_close:
+        if (id == kafka::invalid_fetch_session_id) {
+            reset();
+        } else {
+            close();
+        }
+        return;
+    }
+}
+
+void fetch_session_state::reset() {
+    session_id = kafka::invalid_fetch_session_id;
+    if (_fetch_sessions_enabled()) {
+        session_epoch = kafka::initial_fetch_session_epoch;
+        session_state = state::need_full_fetch;
+    } else {
+        session_epoch = kafka::final_fetch_session_epoch;
+        session_state = state::none;
+    }
+}
+
+void fetch_session_state::toggle(bool enable) {
+    if (enable) {
+        session_state = state::need_full_fetch;
+    } else if (session_id != kafka::invalid_fetch_session_id) {
+        session_state = state::needs_close;
+    } else {
+        session_state = state::none;
+    }
+    update_fetch_session(session_id);
+}
+
 data_queue& fetcher::queue() { return *_parent->_fetched_data_queue; }
 prefix_logger& fetcher::logger() { return _parent->_cluster->logger(); }
 
 fetcher::fetcher(direct_consumer* parent, model::node_id id)
   : _parent(parent)
   , _id(id)
+  , _session_state(_id, logger())
   , _state_lock("fetcher/state") {}
 
 void fetcher::start() {
     ssx::repeat_until_gate_closed(_gate, [this] { return do_fetch(); });
 }
 
 ss::future<> fetcher::stop() {
+    using namespace std::chrono_literals;
     vlog(logger().debug, "[broker: {}] Stopping fetcher", _id);
+    auto fut = co_await ss::coroutine::as_future(ss::with_timeout(
+      model::timeout_clock::now() + 10s, maybe_close_session()));
```

Comment:
after `stop` is called could `fetcher` be destroyed? i ask because `maybe_close_session` would presumably still be running in the background accessing `this`

Reply:
ah right i forgot that with timeout spawns a background fiber. thanks. good eye